File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/K.Рекурсивные_числа_Фибоначчи/solution.py
import sys


memo: dict[int, int] = dict()


def solution(n: int) -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    if n in memo:
        return memo[n]
    if n < 2:
        memo[n] = 1
        return memo[n]
    memo[n] = solution(n - 1) + solution(n - 2)
    return memo[n]


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    sys.stdout.write(str(solution(n)) + '\n')

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/K.Рекурсивные_числа_Фибоначчи/solution.cpp
#include <iostream>
#include <unordered_map>

std::unordered_map<int, int> memo;

int solution(int n) {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    if (memo[n]) {
        return memo[n];
    }
    if (n < 2) {
        memo[n] = 1;
        return memo[n];
    }
    memo[n] = solution(n - 1) + solution(n - 2);
    return memo[n];
}

int main() {
    int n;
    std::cin >> n;
    std::cout << solution(n) << "\n";
}